Add first draft of Popup (Editor's Draft) (#355)
openui/open-ui: research/src/pages/popup/popup.proposal.mdx

---
menu: Proposals
name: Popup (Editor's Draft)
path: /components/popup
pathToResearch: /components/popup.research
---

import '../../styles/spec.css'

> **Editors:** Melanie Richards (Microsoft)

Reviewer: I wouldn't mind being added, if you don't mind.

Author: Perfect, will do!

  </dd>
  <dt>
    <span data-x="concept-element-accessibility-considerations">Accessibility considerations</span>:
  </dt>
  <dd>
    Pending specification in <a href="https://w3c.github.io/html-aria/">ARIA in HTML</a> and{' '}
    <a href="https://w3c.github.io/html-aam/">HTML-AAM</a>.
  </dd>
</dl>

<p class="todo">
  Add IDL to frontmatter (<a href="https://github.com/openui/open-ui/issues/356">#356</a>)
</p>

A `popup` element represents transient application UI that “pops up” in the top layer of a web
document, enabling the user to perform some task in a timely and immediate fashion. `popup` is
similar to the `dialog` element, but is distinguished by its transient nature (“light dismiss”
behaviors) and mutual exclusivity: only one `popup` element may be rendered at a time, with the
exception of nested `popup`s.

<p class="note">
  Examples of popups include: button menus, listboxes used for item selection e.g. comboboxes,
  "teaching" UI, and other progressive-disclosure widgets or content pickers.
</p>

## Showing a `popup` element

A `popup` element should not be shown to the user unless:

- The `initiallyopen` attribute is specified
- The `show()` method is invoked

> - Or the `popup` element has been invoked via the `popup` attribute.

Author: Made a change in 14defbfcc95de7980ff9fd74d439f0df430c5f42

run [showing a `popup` element steps](#showing-popup-steps).

<p class="question">
  When following these run steps, this effectively means the last popup in the DOM order with
  initiallyopen specified will be the popup that "wins". Is that clear enough to implementers? Is
  that a reasonable behavior?
</p>

### The `show()` method

The `show()` method displays the `popup` element, such that the user can interact with it.

**Example:**

```html
<popup id="newFeatureUI">
  <p>
    <strong>New!</strong>
    I’m some sort of educational UI…
  </p>
</popup>

<script>
  if (upsellNewFeature) {
    document.getElementById('newFeatureUI').show()
  }
</script>
```

When `show()` is invoked on a `popup` element _subject_, the user agent must
run [showing a `popup` element steps](#showing-popup-steps).

### The `popup` attribute

The `popup` attribute indicates that a related `popup` element should be shown or hidden as the
user interacts with the element where the `popup` attribute is specified. If the attribute is
specified, the attribute's value must be the ID of a `popup` element in the same
[node document](https://dom.spec.whatwg.org/#concept-node-document).

> **Example:** Invoking the `button` element in this example should show/hide the `popup` element,

Author: Hmm, that makes sense that light dismissal would be what actually hides the popup.

> Click twice, and the popup will still be showing, because it will first light-dismiss, and then the click will cause it to show again.

Could we come up with a way to prevent light dismissal in this case? e.g. don't light dismiss if the event target's attribute has an IDREF relationship with the popup. Otherwise, the `popup` attribute would lead to bad UX on any control that should toggle visibility of the popup, and IMHO is nearly useless.

depending on whether or not the `popup` element is currently shown:

```html
<button id="menuButton" popup="menuPopup">Menu</button>
<popup id="menuPopup" role="menu" anchor="menuButton">
  <!-- Markup for menuitems goes here -->
</popup>
```

The `popup` attribute is supported on a subset of

> [interactive elements](https://html.spec.whatwg.org/multipage/interactive-elements.html):

Author: Any action for me here? I'd say it's intentional that I'm using "interactive elements" in that way that HTML is using it, since I just want to reference that particular list of elements.

> [interactive elements](https://html.spec.whatwg.org/multipage/interactive-elements.html):

Reviewer 2: @melanierichards @mfreed7 only action is to align on definitions outlined here: https://open-ui.org/glossary

The implied definition of Interactive Elements is similar to our Control and that of ARIA's widget. So it's more that we as an industry should be referring to them as the same thing.

- `button`
- `input` in the `button` state

> - `input` in the `text`, `email`, `phone`, or `url` states

Author: Yeah, I think we should limit to this current set of elements to start. The reasoning is not implementation complexity, it's more about which relationships make semantic sense (a commanding element vs a random div that is not perceivable to the user as interactive).

For `date` and `color`, I don't think we need the `popup` attribute because the controller code should take care of the interaction for the popup part.

> - `input` in the `text`, `email`, `phone`, or `url` states

Reviewer 3: +1 to a specified list of options for simplicity. What about `number`, `search`, and `password` types?

> - `input` in the `text`, `email`, `phone`, or `url` states

Author: Thanks @una, I just blanked on a few! `search` makes sense, not sure what the use case would be for `number` but that seems ok. `password` we may want to omit for security purposes but I'm not totally sure on that one.

Let _activating element_ be the element on which the `popup` attribute is specified. If the
_activating element_ is a `button` element or an `input` element in the `button` state, and the

> _activating element_ is activated; OR the _activating element_ is an input in the `text`, `email`,

Author: I think we should either leave this as-is or define some shorthand for the different buckets of valid elements to reference. Reason being, button-y things should show the popup when activated, and other inputs should show the popup when they receive focus. So there is a different logic between the two. Thoughts on which way to resolve?

> `phone`, or `url` states and focus is set to the _activating element_, the user agent must run the

Author: By "either" do you mean "specify only one of these" or "specify that either of these are valid"? Latter makes sense to me.

following steps:

> 1. Let _subject_ be the element with the ID referenced by the `popup` attribute specified on the

Author: These should be all part of the same algo

   _activating element_. If the _subject_ is not a `popup` element, then return.

> 2. If the `open` IDL attribute on the _subject_ is `true`, run

Author: Will leave this one open for now given discussion on the other comment!

> 2. If the `open` IDL attribute on the _subject_ is `true`, run

Author: Added some changes to ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd to address this feedback. Please let me know if the changes resolve the issue or if you have any further comments; feel free to hold until I've resolved all open feedback and have formally requested a new review.

   [hiding a `popup` element steps](#hiding-current-popups).
3. Otherwise, run [showing a `popup` element steps](#showing-popup-steps).

<p class="todo">

>   Probably need to specify some way of storing the relationship once subject is activated, for

Author: Done: #358

  purposes of deciding whether to hide a popup.
</p>

> <p class="question">

Author: Done: #359

  Should we have a way to suppress these steps upon focus, so that the author can enable the popup
  to be shown upon text entry (or some other logic) instead?
</p>

<h3 id="showing-popup-steps">
  Showing a <code>popup</code> element steps
</h3>

1. Run the [hiding currently-shown `popup` elements steps](#hiding-current-popups).
2. Add the _subject_ to _subject_'s
   [node document](https://dom.spec.whatwg.org/#concept-node-document)'s
   [top layer](https://fullscreen.spec.whatwg.org/#top-layer), at the top of the stack.
3. Set the `open` IDL attribute on the _subject_ to `true`.

<h4 id="hiding-current-popups">
  Hiding currently-shown <code>popup</code> elements steps

> </h4>

Author: Please check out the solution in https://github.com/openui/open-ui/pull/355/commits/ecb4964478f3c7330f58914a79955000d906108c and let me know what you think!

Let _candidate subject_ be the new candidate `popup` element to be shown.

> If applicable, let _activating element_ be the element whose `popup` attribute specifies the ID of

Reviewer 3: This will help in the conversations around anchoring (popup invoker/activator <--> popup element)

the _candidate subject_, and which has been activated by the user.
If applicable, let the _anchor element_ be the element whose ID is referenced by the `anchor`
specified on the _candidate subject_.

> Starting at the bottom of the top-layer stack: for each `popup` element which has an `open` IDL

Author: Addressed in https://github.com/openui/open-ui/pull/355/commits/ecb4964478f3c7330f58914a79955000d906108c, please let me know if you have any additional feedback!

attribute set to `true`, and which is not the _candidate subject_:

1. Let _subject_ be the current `popup` element in the top-layer stack.
2. Walk the DOM to determine if the _subject_ is an ancestor to the _candidate subject_, the
   _activating element_, or the _anchor element_.
3. If the _subject_ is an ancestor, then return.

> 4. Otherwise, for this _subject_ and any other `popup`s higher in the stack, run the

Author: Made an adjustment in https://github.com/openui/open-ui/pull/355/commits/5e04a4e6435d958690227484ac5dbd073233b5a6. I imagine that text can be workshopped so let me know if you have any additional comments!

   [hiding a `popup` element steps](#hiding-current-popups).

<p class="todo">

>   Step #2 probably needs to be more precise, and more specific about Shadow/light DOM

Reviewer: Step 2 should definitely refer to the "flat tree" (defined in CSS [here](https://drafts.csswg.org/css-scoping/#flattening)). See Chromium code [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/html_popup_element.cc;l=265;drc=1de0bd095828489ebfef0c265e172af3d1dd0afd). This was important to enable the `<selectmenu>`'s use of shadow DOM.

</p>

## Setting focus when the `popup` element is shown

### The `delegatesfocus` attribute

The `delegatesfocus` attribute is a
[boolean attribute](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attribute).

> When specified, it indicates that when the `popup` element is shown, focus should move to the first

Reviewer: If you're referring to [this version of delegatesfocus](https://dom.spec.whatwg.org/#dom-shadowroot-delegatesfocus), that's fairly different. That refers to Shadow DOM and how focus is proxied into the shadow root (and isn't well specified, at least in DOM/HTML specs). I bring this up for two reasons:
 1. I do think this version here needs a specific definition like this, to be clear about what it does.
 2. I am wondering if we should overload this attribute name for two similar but different concepts. Not sure if it helps or hurts developer understandability.

> When specified, it indicates that when the `popup` element is shown, focus should move to the first

Author: +1 to this being a net-new attribute, even if the name was inspired from the attr in Shadow DOM. @mfreed7, are there any specific changes you'd want to see in the current definition here?

Also, I opened #367 and #368 to capture some of the other points here that I think we can defer until after merging this PR.

> When specified, it indicates that when the `popup` element is shown, focus should move to the first

Reviewer 2: Yeah, I wasn't aware they were different and so I do think that there will be confusion. I'd be curious how "different" the concepts are prior to agreeing on overloading the name. I am fine with having the steps based on @mfreed7 feedback. @melanierichards feel free to resolve this after you file an issue regarding the overloading of the attribute

focusable descendent of the `popup` element.

**Example:**

```html
<popup delegatesfocus>
  <p>I am not a focusable element.</p>
  <p>Nor am I.</p>
  <button>I will be focused whenever the popup becomes focused.</button>
</popup>
```

### Focusing steps

The popup [focusing steps](https://html.spec.whatwg.org/multipage/interaction.html#focusing-steps)
for a `popup` element _subject_ are as follows:

> 1. If the `delegatesfocus` attribute is specified on _subject_, let _control_ be the first focusable

Author: See https://github.com/openui/open-ui/pull/355/files#r652867167

>    descendant element of _subject_, in [tree order](https://dom.spec.whatwg.org/#concept-tree-order),

Reviewer: So the spec for ["Get the focusable area"](https://html.spec.whatwg.org/multipage/interaction.html#get-the-focusable-area) should be used here, I think. And actually, I think it'd be a lot better if we folded this algorithm (focusing steps for `<popup>` and even `<dialog>`) back into the "Get the focusable area" spec. It'd be a lot cleaner, I think.

Author: Done in https://github.com/openui/open-ui/pull/355/commits/ecb4964478f3c7330f58914a79955000d906108c! I've created a new section of this document at the bottom for text that should land in the focusable area algorithm.

   that is not inert and has the

>    [autofocus](https://html.spec.whatwg.org/multipage/interaction.html#attr-fe-autofocus) attribute

Reviewer: This is a bit different than I [implemented](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/html_popup_element.cc;l=155;drc=1de0bd095828489ebfef0c265e172af3d1dd0afd) in Chromium, in that the `delegatesfocus` attribute causes the first focusable element to be focused, regardless of whether it **also** has the `autofocus` attribute. If `subject` does **not** have the `delegatesfocus` attribute, **then** the code looks for a descendant element that has the `autofocus` attribute, and focuses that.

I can change that behavior, but it actually makes a bit more sense to me than the algorithm here, as written. As written, you need to **both** add `delegatesfocus` **and** add `autofocus` to an element, which means there's no way to just say "focus the first thing you can within this popup". It feels like these should be two separate features:
 1. `delegatesfocus` on the `<popup>` itself: focus the first focusable thing inside
 2. `autofocus` on the `<popup>` or any contained element: focus that

Author: Honestly, I think I just missed moving some text around, because you're right, the algo as written makes no sense. Sorry about that and thank you for flagging it! Should be fixed in this commit: https://github.com/openui/open-ui/pull/355/commits/252119e3d1274313d70f6bf4fb5b21ecfb0d587b

   specified.

> 2. If there is no such descendent, but `delegatesfocus` is specified, let _control_ be _subject_.

Reviewer 2: Also not sure if we should use the term _control_ here, possibly use a different variable name

> 2. If there is no such descendent, but `delegatesfocus` is specified, let _control_ be _subject_.

Author: For the first comment, refer to https://github.com/openui/open-ui/pull/355/files#r652867167.

Re _control_: yeah, it's a bit overloaded. I'm basically adopting WHATWG spec language part and parcel on this one. Any suggestions? Or maybe we just leave the host spec language for now.

> 2. If there is no such descendent, but `delegatesfocus` is specified, let _control_ be _subject_.

Author: This is now moot, as I'm using _focus target_. Closing!

> 3. Otherwise, if the `autofocus` attribute is specified on _subject_, let _control_ be _subject_.

Reviewer: This is, unfortunately, how WHATWG typically writes if/else algorithms. Step N is "if X, do Y", and step N+1 is "otherwise, do Z".

> 3. Otherwise, if the `autofocus` attribute is specified on _subject_, let _control_ be _subject_.

Author: +1, I elected to follow WHATWG style here. Inviting HTML editors to weigh in on whether they'd be amenable to sub-bullets or if we should follow house style in order to land this text with minimal changes.

> 4. Otherwise, return.

Reviewer: returning from the algorithm.

Here, I'm not sure about WHATWG conventions. This is an if/elseif/else block, as step 4 is the "otherwise" for the embedded "if" in step 3. That, I haven't exactly seen.

> 4. Otherwise, return.

Author: Any changes requested, e.g. melding 4 into step 3?

   [top-level browsing context](https://html.spec.whatwg.org/multipage/browsers.html#top-level-browsing-context).
7. If control's [node document](https://dom.spec.whatwg.org/#concept-node-document)'s
   [origin](https://dom.spec.whatwg.org/#concept-document-origin) is not the
   [same](https://html.spec.whatwg.org/multipage/origin.html#same-origin) as the
   [origin](https://dom.spec.whatwg.org/#concept-document-origin) of _topDocument_, then return.
8. [Empty](https://infra.spec.whatwg.org/#list-empty) _topDocument_'s
   [autofocus candidates](https://html.spec.whatwg.org/multipage/interaction.html#autofocus-candidates).
9. Set _topDocument_'s
   [autofocus processed flag](https://html.spec.whatwg.org/multipage/interaction.html#autofocus-processed-flag)
   to true.

<p class="note">
  Without the precense of the autofocus or delegatefocus attribute, focus remains on the active
  element. This behavior is to enable scenarios where the popup is used in a composite control. For
  example, a combobox where the user expects their focus to stay in the text input instead of moving
  automatically to the listbox popup as and when it appears.
</p>

> <p class="question">

Reviewer 2: Open issue

Author: Done: #369

>   These focusing steps effectively assert that the delegatesfocus attribute takes precedence over

Reviewer: That's also how I implemented it, and it made sense to me at the time. For whatever that's worth! 😆

Author: Huzzah!

[node document](https://dom.spec.whatwg.org/#concept-node-document) as the `popup` element.

**Example:**

```html
<button id="myButton">Anchor element</button>
<popup open anchor="myButton">
  <p>
    <strong>New!</strong>
    I’m some sort of educational UI…
  </p>
</popup>
```

The target of the `anchor` attribute participates in determining whether a candiate _popup_ element
is "nested" in a currently-shown `popup`; refer to <a href="#hiding-current-popups">hiding
currently-shown <code>popup</code> elements steps</a>.

<p class="todo">
  Whenever there is a spec to point to for CSS anchored positioning, call out here that the anchor
  attribute is used to help resolve CSS <code>anchor()</code> functions.
</p>

## Hiding a `popup` element

### The `hide()` method

The `hide()` method hides a `popup` element. When `hide()` is invoked, the user agent must run
[hiding a `popup` element steps](#hiding-popup-steps) from the `popup` element _subject_ on which
the method was invoked.

### Light dismissal

When a `popup` element is shown and a light dismiss interaction occurs, the user agent must run

> [hiding a `popup` element steps](#hiding-popup-steps) from the `popup` element _subject_.

Reviewer: Here is what I was referring to earlier. When light-dismissing, there will be a stack of open popups, and the light dismiss event doesn't have context to **which** popup should be `subject`. It will, however, have a target Node (in the case of a "click outside"). That Node should be the parameter used with the (new) hiding a popup element `start node` parameter. In code, for a "user presses ESCAPE" type light dismiss, I actually defined another algorithm to "hide the top popup" only. And for a "user scrolls the page" or "resizes the window" or other such light dismiss actions, you would call "hide the popup" with `start node` set to `null`, indicating that all open popups should be closed.

Author: See above comment about solution: https://github.com/openui/open-ui/pull/355/commits/ecb4964478f3c7330f58914a79955000d906108c

<p class="todo">
  Land <a href="https://github.com/openui/open-ui/pull/243">#243</a> and link out to this canonical
  definition of light dismissal for consistency. Layout changes to the popup or its anchor element
  are also ground for light dismissal, figure out how to surface this special case. How to handle in

>   prose light dismiss behaviors that may dismiss some but not all currently-shown popups?

Reviewer: See comment above for this.

Author: See above comment about solution: https://github.com/openui/open-ui/pull/355/commits/ecb4964478f3c7330f58914a79955000d906108c

</p>

> <h3 id="hiding-popup-steps">

Author: I'm of two minds about this, because I think what you're saying does make sense. But on the other hand, I've tried to order these sections by popup lifecycle (showing, focusing, anchoring, hiding) and it might read a little funny if those are out of order. The "Hiding other popups" section probably makes that a little less clear, since it is about _hiding_ (end of lifecycle) but is a subtask of showing algos. Thoughts?

> <h3 id="hiding-popup-steps">

Author: I ended up moving things around like this: https://github.com/openui/open-ui/pull/355/commits/ecb4964478f3c7330f58914a79955000d906108c

  </dd>
  <dt>
    <span data-x="concept-element-accessibility-considerations">Accessibility considerations</span>:
  </dt>
  <dd>
    Pending specification in <a href="https://w3c.github.io/html-aria/">ARIA in HTML</a> and{' '}
    <a href="https://w3c.github.io/html-aam/">HTML-AAM</a>.
  </dd>
</dl>

<p class="todo">
  Add IDL to frontmatter (<a href="https://github.com/openui/open-ui/issues/356">#356</a>)
</p>

A `popup` element represents transient application UI that “pops up” in the
[top layer](https://fullscreen.spec.whatwg.org/#top-layer) of a web document, enabling the user to
perform some task in a timely and immediate fashion. `popup` is similar to the `dialog` element, but
is distinguished by its transient nature (“light dismiss” behaviors) and mutual exclusivity: only
one `popup` element may be rendered at a time, with the exception of nested `popup`s.

<p class="note">
  Examples of popups include: button menus, listboxes used for item selection e.g. comboboxes,
  "teaching" UI, and other progressive-disclosure widgets or content pickers.
</p>

## Showing a `popup` element

A `popup` element and its contents should not be rendered unless:

> - The `initiallyopen` attribute is specified

Reviewer 3: This might be bike-sheddy so feel free to ignore for now, but should it be `initiallyOpen`?

Author: Since attributes are case-insensitive, either will work. Camel-casing seems easier to read but looks like the HTML spec has a stylistic preference for all lowercase. Will let HTML editors chime in here!

implement this requirement using the CSS rules described in the
[Rendering section](https://html.spec.whatwg.org/multipage/rendering.html#rendering).

<h3 id="popup-stack">The popup stack</h3>

This specification introduces a new stacking layer in the document
[top layer](https://www.w3.org/TR/CSS2/zindex.html), called the **popup stack**. Each document has
one popup stack.

The popup stack is an [ordered set](https://infra.spec.whatwg.org/#ordered-set) of elements,
rendered in the order they appear in the set. The last element in the set is rendered last,
and thus appears on top.

Showing or hiding a `popup` element adds or removes the element from the popup stack, respectively.

### The `initiallyopen` attribute

The `initiallyopen` attribute is a
[boolean attribute](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attribute).
When specified, it indicates that the `popup` element should have the `show()` method called upon
document load, such that the user can interact with it.

**Example:**

```html
<popup initiallyopen></popup>
```

If the `initiallyopen` attribute is specified on a `popup` element _subject_, the user agent must

> run [showing a `popup` element steps](#showing-popup-steps).

Author: Added some changes to ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd to address this feedback (including higher up in the document). Please let me know if the changes resolve the issue or if you have any further comments; feel free to hold until I've resolved all open feedback and have formally requested a new review.

<p class="note">
  As a result of following these run steps: in the event that initiallyopen is specified on multiple
  popups, the last of these popups (with the initiallyopen attribute) in the DOM order will be the
  popup that is rendered upon document load.
</p>

### The `show()` method

The `show()` method renders the `popup` element, such that the user can interact with it.

**Example:**

```html
<popup id="newFeatureUI">
  <p>
    <strong>New!</strong>
    I’m some sort of educational UI…
  </p>
</popup>

<script>
  if (upsellNewFeature) {
    document.getElementById('newFeatureUI').show()
  }
</script>
```

When `show()` is invoked on a `popup` element _subject_, the user agent must

> run [showing a `popup` element steps](#showing-popup-steps).

Author: Added some changes to ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd to address this feedback. Please let me know if the changes resolve the issue or if you have any further comments; feel free to hold until I've resolved all open feedback and have formally requested a new review.

- `input` in the `email`, `number`, `search`, `tel`, `text`, or `url` states

#### Showing/hiding a `popup` element via the `popup` attribute

When an element with the `popup` attribute specified receives a user interaction that should show
the related `popup` element, this element shall be known as the `popup` element's _invoker_.

<p class="note">
  Multiple elements may have a popup attribute that refers to the same popup element. A given popup
  element only has 0 or 1 invokers at a given time. For example, if the popup element was shown
  programmatically, the invoker is null. Otherwise, the element that the user actually interacted
  with to show the popup is its sole invoker.
</p>

If the _invoker_ is a `button` element or an `input` element in the `button` state, and the
_invoker_ is activated; OR the _invoker_ is an input in the `email`, `number`, `search`, `tel`,
`text`, or `url` states and focus is set to the _invoker_, the user agent must run the following
steps:

1. Let _subject_ be the element with the ID referenced by the `popup` attribute specified on the
   _invoker_. If there is no such element, or the _subject_ is not a `popup` element,

>    then return.

Author: Good catch!

Added some changes to ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd to address this feedback. Please let me know if the changes resolve the issue or if you have any further comments; feel free to hold until I've resolved all open feedback and have formally requested a new review.

2. If the `open` IDL attribute on the _subject_ is `true`, run
   [hiding a `popup` element steps](#hiding-current-popups).
3. Otherwise, run [showing a `popup` element steps](#showing-popup-steps).

<p class="question">
  Should we have a way to suppress these steps upon focus, so that the author can enable the popup
  to be shown upon text entry (or some other logic) instead?
</p>

<h3 id="showing-popup-steps">
  Showing a <code>popup</code> element steps
</h3>

Let _candidate subject_ be the new candidate `popup` element to be shown.
If applicable, let _invoker_ be the element whose `popup` attribute specifies the ID of
the _candidate subject_, and which has been activated by the user.
If applicable, let the _anchor element_ be the element whose ID is referenced by the `anchor`

> specified on the _candidate subject_.

Author: Added some changes to ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd to address this feedback. Please let me know if the changes resolve the issue or if you have any further comments; feel free to hold until I've resolved all open feedback and have formally requested a new review.

1. Run the [hiding currently-shown `popup` elements steps](#hiding-current-popups).
2. Add the _subject_ to _subject_'s
   [node document](https://dom.spec.whatwg.org/#concept-node-document)'s [popup stack](#popup-stack).
3. Set the `open` IDL attribute on the _subject_ to `true`.
4. If the `popup` element was shown as a result of a user interaction with an `invoker`, internally

>    set the _subject_'s _invoker_ to a pointer to the `invoker` element.

Author: Can you help me better understand your feedback on this one? Per your feedback on line 216, we'd already be passing in the _invoker_ parameter, we just need to make sure that relationship is stored elsewhere so that it can be referenced later by hide steps...would it be preferable to provide that directive where "showing a popup element steps" are called, vs as a step in the algorithm itself?

>    set the _subject_'s _invoker_ to a pointer to the `invoker` element.

Reviewer 4: Sure. The idea is that instead of "If the `popup` element was shown as a result of a user interaction with an `invoker`", this should be something like "If _dueToUserInteraction_ is true". Then, all call sites of the "showing a popup element steps" algorithm either pass true or false for that parameter.

>    set the _subject_'s _invoker_ to a pointer to the `invoker` element.

Author: Thanks Domenic! Another way perhaps to solve this is to say "If _invoker_ is not null, internally set the _candidate subject_'s _invoker_ to a pointer to the `invoker` element." In this case, per the changes in ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd, we've specified that these steps take an optional _invoker_ param, and all relevant call sites pass that parameter. Thoughts on that vs. just adding another paremeter?

> 5. Run [focusing steps](#focusing-steps).

Author: Added some changes to ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd to address this feedback. Please let me know if the changes resolve the issue or if you have any further comments; feel free to hold until I've resolved all open feedback and have formally requested a new review.

> ## Setting focus when the `popup` element is shown

Author: Ok, thanks! We can strip them out to fit house style once we've incubated this feature a bit more and are ready to open a PR against HTML.

When specified on the `popup` element, it indicates that when the `popup` element is shown, focus
should move to the `popup` element.

**Example:**

```html
<popup autofocus>
  <p>The popup itself will receive focus.</p>
  <button>A focusable element</p>
</popup>
```

When specified on a descendent of the `popup` element, it indicates that when the `popup` element is
shown, focus should move to the descendent of the `popup` element where `autofocus` is specified.

**Example:**

```html
<popup>
  <p>Text goes here</p>
  <button>Button one</button>
  <button autofocus>Button two</p>
</popup>
```

### The `delegatesfocus` attribute

The `delegatesfocus` content attribute is a
[boolean attribute](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attribute).
When specified, it indicates that when the `popup` element is shown, focus should move to the first
focusable descendent of the `popup` element.

**Example:**

```html
<popup delegatesfocus>
  <p>I am not a focusable element.</p>
  <p>Nor am I.</p>
  <button>I will be focused whenever the popup becomes focused.</button>
</popup>
```

<h3 id="focusing-steps">Focusing steps</h3>

> When a `popup` is shown, user agents must set the candidate `popup` element to the _focus target_,

Author: Added some changes to ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd to address this feedback. Please let me know if the changes resolve the issue or if you have any further comments; feel free to hold until I've resolved all open feedback and have formally requested a new review.

then run [get the focusable area
steps](https://html.spec.whatwg.org/multipage/interaction.html#get-the-focusable-area) and

> [focusing steps](https://html.spec.whatwg.org/multipage/interaction.html#focusing=steps).

Author: Added some changes to ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd to address this feedback. Please let me know if the changes resolve the issue or if you have any further comments; feel free to hold until I've resolved all open feedback and have formally requested a new review.

**Example:**

```html
<button id="myButton">Anchor element</button>
<popup open anchor="myButton">
  <p>
    <strong>New!</strong>
    I’m some sort of educational UI…
  </p>
</popup>
```

The target of the `anchor` attribute participates in determining whether a candiate _popup_ element
is "nested" in a currently-shown `popup`; refer to <a href="#hiding-current-popups">hiding
currently-shown <code>popup</code> elements steps</a>.

<p class="todo">
  Whenever there is a spec to point to for CSS anchored positioning, call out here that the anchor
  attribute is used to help resolve CSS <code>anchor()</code> functions.
</p>

## Hiding a `popup` element

### The `hide()` method

The `hide()` method hides a `popup` element. When `hide()` is invoked, the user agent must run
[hiding a `popup` element steps](#hiding-popup-steps) from the `popup` element _subject_ on which
the method was invoked.

### Light dismissal

"Light dismiss" describes the behavior of a component which dismisses itself automatically when a
user performs an action that implies that they are finished interacting with the component. The
actions that trigger light dismiss are:

- The user presses the escape key.

> - A focus change occurs (because of either user interaction or script), where the focus target is

Author: Thanks Domenic, I'm trying to square this comment with your [feedback here](https://github.com/openui/open-ui/pull/355/files#r675188384). Do we want focus-oriented "light dismissal" triggers to live in popup's close watcher logic, in the focusing steps algo, or both?

> - A focus change occurs (because of either user interaction or script), where the focus target is

Author: See also my [comment here on a desire to componentize a light dismissal definition](https://github.com/openui/open-ui/pull/355/files#r684443042), if that influences the direction any.

> - A focus change occurs (because of either user interaction or script), where the focus target is

Reviewer 4: Well, I think both. As far as I can tell you want to close the popup in two situations:

- When a close signal comes in (e.g. Esc key, or back button on Android)
- When a focus change occurs

Each of them requires spec patches.

As for "componentizing" light dismissal, I would encourage not creating abstractions that only have one consumer, but instead waiting to do that until multiple parts of the platform use it.

> - A focus change occurs (because of either user interaction or script), where the focus target is

Author: Got it, thanks!

  outside of the contents of the component. This includes the case where the user invokes a
  non-focusable element, which causes focus to jump to the nearest focusable ancestor of that element.
- An OS-level focus change occurs such that the window containing the component no longer has focus.
  For example, the user switches to a different browser tab or switches to a different application.

<p class="todo">
  Land <a href="https://github.com/openui/open-ui/pull/243">#243</a> and link out to this canonical

>   definition of light dismissal for consistency (remove inline definition). Layout changes to the

Author: Yes sorry, intent is to componentize the definition of light dismissal somewhere. First "canonical" definition will live in Open UI until it's sufficiently incubated to land in HTML, which would then become canon; everything (including `popup` in HTML) would link to that light dismissal definition in HTML.

>   definition of light dismissal for consistency (remove inline definition). Layout changes to the

Author: Addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

  popup or its anchor element are also ground for light dismissal, figure out how to surface this
  special case.
</p>

When a `popup` element is shown and a light dismiss interaction occurs:

1. If a keypress event for the `ESCAPE` key was fired, the user agent must set the _subject_ to
   the `popup` element and run [hiding a `popup` element steps](#hiding-popup-steps).
2. If a focus change occurs, the user agent must set the _start node_ to the element which received
   the focus event and run [Hiding currently-shown `popup` elements steps](#hiding-current-popups).
3. Otherwise, the user agent must set the _start node_ to null and run [Hiding currently-shown
   `popup` elements steps](#hiding-current-popups).

<h3 id="hiding-current-popups">
  Hiding currently-shown <code>popup</code> elements steps
</h3>

> Starting at the top of the [popup stack](#popup-stack):

Reviewer 4: It's not clear what this means.

Author: Can you help me understand what's unclear?

Reviewer 4: Well, first, do stacks grow up or down? :) It's not clear which is the top.

Second, which popup stack? This algorithm probably needs to take a document as an input, so you can look at that document's popup stack in particular.

Author: Cool, added a reference to the popup's node document in ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd, but I'll see what else I can do in the popup-stack section to clarify.

> 1. Let _subject_ be the current `popup` element in the [popup stack](#popup-stack).

Reviewer 4: Do you want to ever remove it from the stack? Probably using [pop](https://infra.spec.whatwg.org/#stack-pop)? As written I believe it is left in the stack.

Author: Step 6 calls out to an algorithm that would remove the subject from the stack (although I did need to make an update to make that clearer in ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd). Note to self: update language from "remove" => "pop" to better fit house style.

> 2. If _candidate subject_ is null, skip to step 6.

Author: Added some changes to ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd to address this feedback. Please let me know if the changes resolve the issue or if you have any further comments; feel free to hold until I've resolved all open feedback and have formally requested a new review.

3. Otherwise, starting with the _candidate subject_, walk the
   [flat tree](https://drafts.csswg.org/css-scoping/#flattening) to determine whether the _subject_ is

>    an ancestor to the _candidate subject_. If the _subject_ is an ancestor, then return.

Reviewer 4: I think I might be wrong about this comment. Although flat tree doesn't seem to be well-defined to the same extent as other DOM relationships, a number of specs seem to be based on top of it and implemented without ambiguity. @mfreed7 might be able to comment better.

>    an ancestor to the _candidate subject_. If the _subject_ is an ancestor, then return.

Author: Cool, I'll hold on making changes on this one then!

4. Otherwise, if the _invoker_ is not null, starting with the _invoker_, walk the
   [flat tree](https://drafts.csswg.org/css-scoping/#flattening) to determine whether the _subject_ is
   an ancestor to the _invoker_. If the _subject_ is an ancestor, then return.
5. Otherwise, if the _anchor element_ is not null, starting with the _anchor element_, walk the
   [flat tree](https://drafts.csswg.org/css-scoping/#flattening) to determine whether the _subject_ is
   an ancestor to the _anchor element_. If the _subject_ is an ancestor, then return.
6. If _start node_ is not null, starting with the _start node_, walk the
   [flat tree](https://drafts.csswg.org/css-scoping/#flattening) to determine whether the _subject_ is
   the _start node_ or an ancestor to the _start node_. If either condition is met, then return.
7. Otherwise, for this _subject_, run the
   [hiding a `popup` element steps](#hiding-popup-steps). Repeat for the next `popup` element in

>    the [popup stack](#popup-stack).

Author: Added some changes to ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd to address this feedback. Please let me know if the changes resolve the issue or if you have any further comments; feel free to hold until I've resolved all open feedback and have formally requested a new review.

<h3 id="hiding-popup-steps">
  Hiding a <code>popup</code> element steps
</h3>

The hiding steps for a `popup` element _subject_ are as follows:

1. Remove the _subject_ from the
   [node document](https://dom.spec.whatwg.org/#concept-node-document)'s
   [top layer](https://fullscreen.spec.whatwg.org/#top-layer).
2. Set the `open` IDL attribute on the _subject_ to `false`.

> 3. If applicable, empty the _subject_'s _invoker_.

Reviewer 4: _invoker_ is not a declared variable... Maybe this is supposed to be a link to a definition?

Author: Added some changes to ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd to address this feedback. Please let me know if the changes resolve the issue or if you have any further comments; feel free to hold until I've resolved all open feedback and have formally requested a new review.

4. [Queue an element task](https://html.spec.whatwg.org/multipage/webappapis.html#queue-an-element-task)
   on the [user interaction task source](https://html.spec.whatwg.org/multipage/webappapis.html#user-interaction-task-source)

>    given the _subject_ element to [fire a non-cancelable event](https://dom.spec.whatwg.org/#concept-event-fire)

Reviewer 4: ```suggestion
   given the _subject_ element to [fire an event](https://dom.spec.whatwg.org/#concept-event-fire)
```

Author: Added some changes to ccd9b75036cb67e177ad11b2bf72b0b9c1ed0afd to address this feedback. Please let me know if the changes resolve the issue or if you have any further comments; feel free to hold until I've resolved all open feedback and have formally requested a new review.

    the popup is shown
  </dd>
  <dd>
    <code>initiallyopen</code> — Whether the popup is shown upon document load
  </dd>
  <dt>
    <span data-x="concept-element-accessibility-considerations">Accessibility considerations</span>:
  </dt>
  <dd>
    Pending specification in <a href="https://w3c.github.io/html-aria/">ARIA in HTML</a> and{' '}
    <a href="https://w3c.github.io/html-aam/">HTML-AAM</a>.
  </dd>
</dl>

<p class="todo">
  Add IDL to frontmatter (<a href="https://github.com/openui/open-ui/issues/356">#356</a>)
</p>

A `popup` element represents transient application UI that “pops up” in the
[top layer](https://fullscreen.spec.whatwg.org/#top-layer) of a web document, enabling the user to
perform some task in a timely and immediate fashion. `popup` is similar to the `dialog` element, but
is distinguished by its transient nature (“light dismiss” behaviors) and mutual exclusivity: only
one `popup` element may be rendered at a time, with the exception of nested `popup`s.

When a `popup` is
[browsing-context connected](https://html.spec.whatwg.org/#becomes-browsing-context-connected), set
its _open state_ to `false`, prior to running any

> [showing a `popup` element steps](#showing-popup-steps) on the element, as applicable.

Author: I suppose it's the latter, in which case a discussion on the correct resolution would be helpful. I'm thinking that this verbiage here can be dropped, and then a step 2 would be added to this algo: https://github.com/openui/open-ui/pull/355/files#diff-21cce170c953f236921ecbc4481a202782c1b2bc7464307c76aaa4dc60b64f11R131

The effect (paraphrasing in prose) would be: when the popup is browsing-context-connected, check for the `initiallyopen` content attribute. If present, set this open state to true and run through popup-showing steps. Else, set the open state to false. Would that change resolve the issue?

> [showing a `popup` element steps](#showing-popup-steps) on the element, as applicable.

Author: Ended up doing something like this in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

<p class="note">
  Examples of popups include: button menus, listboxes used for item selection e.g. comboboxes,
  "teaching" UI, and other progressive-disclosure widgets or content pickers.
</p>

## Showing a `popup` element

> A `popup` element and its contents should not be rendered unless:

Author: Must! Addressed with 5fc5041409d28c395e81cf30f039a7c0e269b7fe

This requirement may be implemented indirectly through the style layer. For example,
user agents that
[support the suggested default rendering](https://html.spec.whatwg.org/multipage/infrastructure.html#renderingUA)
implement this requirement using the CSS rules described in the
[Rendering section](https://html.spec.whatwg.org/multipage/rendering.html#rendering).

<h3 id="popup-stack">The popup stack</h3>

This specification introduces a new stacking layer in the document
[top layer](https://www.w3.org/TR/CSS2/zindex.html), called the **popup stack**. Each document has
one popup stack.

The popup stack is an [ordered set](https://infra.spec.whatwg.org/#ordered-set) of elements,
rendered in the order they appear in the set. The last element in the set is rendered last,
and thus appears on top.

Showing or hiding a `popup` element adds or removes the element from the popup stack, respectively.

### The `initiallyopen` attribute

The `initiallyopen` attribute is a
[boolean attribute](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attribute).
When specified, it indicates that the `popup` element should have the `show()` method called upon

> document load, such that the user can interact with it.

Author: Thanks, I'm not sure why I missed these shoulds. Perhaps because not written as SHOULD? :)

Anyway, addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

**Example:**

```html
<popup initiallyopen></popup>
```

When a `popup` element becomes
[browsing-context connected](https://html.spec.whatwg.org/#becomes-browsing-context-connected), the
user agent must run the following steps:

1. If the `initiallyopen` content attribute is specified on the `popup` element, run the
   [showing a `popup` element steps](#showing-popup-steps) with _candidate subject_ set to the current
   `popup` element.

> 2. Otherwise, return.

Author: Addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

</popup>
```

The `popup` attribute is supported on a subset of
[interactive elements](https://html.spec.whatwg.org/multipage/interactive-elements.html):

- `button`
- `input` in the `button` state
- `input` in the `email`, `number`, `search`, `tel`, `text`, or `url` states

#### Showing/hiding a `popup` element via the `popup` attribute

When an element with the `popup` attribute specified receives a user interaction that should show
the related `popup` element, this element shall be known as the `popup` element's _invoker_.

<p class="note">
  Multiple elements may have a popup attribute that refers to the same popup element. A given popup
  element only has 0 or 1 invokers at a given time. For example, if the popup element was shown
  programmatically, the invoker is null. Otherwise, the element that the user actually interacted
  with to show the popup is its sole invoker.
</p>

If the _invoker_ is a `button` element or an `input` element in the `button` state, and the
_invoker_ is activated; OR the _invoker_ is an input in the `email`, `number`, `search`, `tel`,
`text`, or `url` states and focus is set to the _invoker_, the user agent must run the following

> steps:

Author: Sounds good. Couple things I'm not sure of:

1. We'd have to run these steps from 8 different places (element types + applicable states) across the HTML spec. Should the steps live in one central location that we can call out to from each applicable `button`/`input` state? Or do we prefer to repeat everywhere they should appear?
2. Just a heads up that I think it would be good to have some note here pointing to the various different places that we move these steps, so that implementers/authors/other audiences can cross-reference from `popup`. Happy to take suggestions on how that's done.

Also, I don't think there's any effect on the focus update steps from this piece directly, since we are persisting user focus (unless moved via the `autofocus` or `delegatesfocus` attribute).

1. Let _subject_ be the first element in tree order with the ID referenced by the `popup` attribute
   specified on the _invoker_. If there is no such element, or the _subject_ is not a `popup`
   element, then return.
2. Get the _subject_'s _open state_.
3. If _subject_'s _open state_ is `true`, run

>    [hiding a `popup` element steps](#hiding-current-popups).

Author: Addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

4. Otherwise, run [showing a `popup` element steps](#showing-popup-steps), with _candidate subject_
   set to _subject_ and _invoker_ set to the current element _invoker_.

<p class="question">
  Should we have a way to suppress these steps upon focus, so that the author can enable the popup
  to be shown upon text entry (or some other logic) instead?
</p>

<h3 id="showing-popup-steps">
  Showing a <code>popup</code> element steps
</h3>

The **showing a `popup` element steps**, given a popup element _candidate subject_ and an optional
element _invoker_, and an optional element _anchor element_, are:

1. Run the [hiding currently-shown `popup` elements steps](#hiding-current-popups), with the
   _candidate subject_, optional element _invoker_, and optional element _anchor element_.
2. Add the _candidate subject_ to _candidate subject_'s
   [node document](https://dom.spec.whatwg.org/#concept-node-document)'s [popup stack](#popup-stack).

> 3. Set the `open` IDL attribute on the _candidate subject_ to `true`.

Author: I added step 4 here based on the previous feedback to hold this in an internal state mechanism, so satisfying this comment would be a case of removing (3).

What is a bit unclear to me is why we wouldn't want to have the IDL attr reflect the UA-known state of the popup. Couldn't that create clashes in the other direction, because the state is out of sync/inaccurate?

> 3. Set the `open` IDL attribute on the _candidate subject_ to `true`.

Reviewer 4: I think the IDL attribute should reflect the internal state. I expect its spec to be "The open getter steps are to return this's open state" (or whatever you named the internal state; I haven't looked at the diff yet :)).

> 3. Set the `open` IDL attribute on the _candidate subject_ to `true`.

Author: Ah, I better understand the feedback now. Thanks!

> 3. Set the `open` IDL attribute on the _candidate subject_ to `true`.

Author: Addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

4. Set the _candidate subject_'s _open state_ to `true`.
5. If the `popup` element was shown as a result of a user interaction with an `invoker`, internally

>    set the _candidate subject_'s _invoker_ to a pointer to the `invoker` element.

Author: Addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

6. Run [get the focusable area
   steps](https://html.spec.whatwg.org/multipage/interaction.html#get-the-focusable-area) with
   _focus target_ set to the _candidate subject_.
7. Run [focusing steps](https://html.spec.whatwg.org/multipage/interaction.html#focusing=steps)
   with _new focus target_ set to the returned node from [get the focusable area
   steps](https://html.spec.whatwg.org/multipage/interaction.html#get-the-focusable-area).

<p class="note">

>   Without the precense of the autofocus or delegatefocus attribute, focus remains on the active

Author: You've found my spelling kryptonite! Let me look into the other note.

  element. This behavior is to enable scenarios where the popup is used in a composite control. For
  example, a combobox where the user expects their focus to stay in the text input instead of moving
  automatically to the listbox popup as and when it appears.
</p>

## Setting initial focus with the `autofocus` attribute

The [autofocus](https://html.spec.whatwg.org/multipage/interaction.html#the-autofocus-attribute)
content attribute allows the author to indicate that an element is to be focused as soon as the
page is loaded or as soon as the `popup` within which it finds itself is shown.

When specified on the `popup` element, it indicates that when the `popup` element is shown, focus

> should move to the `popup` element.

Author: Addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

  special case.
</p>

When a `popup` element is shown and a light dismiss interaction occurs:

1. If a keypress event for the `ESCAPE` key was fired, run [hiding a `popup` element
   steps](#hiding-popup-steps), with the _candidate subject_ set to the top-most `popup` element in
   the [popup stack](#pop-stack).
2. If a focus change occurs, the user agent must set the _start node_ to the element which received
   the focus event and run [Hiding currently-shown `popup` elements steps](#hiding-current-popups).
3. Otherwise, the user agent must set the _start node_ to null and run [Hiding currently-shown
   `popup` elements steps](#hiding-current-popups).

<h3 id="hiding-current-popups">
  Hiding currently-shown <code>popup</code> elements steps
</h3>

The **hiding currently-shown `popup` elements steps**, given an optional popup element
_candidate subject_, an optional element _invoker_, and an optional element _anchor element_, and an
optional element _start node_, are as follows.

For each popup element _subject_ in the
[node document](https://dom.spec.whatwg.org/#concept-node-document)'s [popup stack](#popup-stack),

> from top to bottom:

Reviewer 4: Still unclear what "top to bottom" means. ([Do stacks grow up or down?](https://stackoverflow.com/questions/1677415/does-stack-grow-upward-or-downward))

For the purposes of "For each", stacks in web standards are ordered by insertion order. So "For each _subject_ of the popup stack" would iterate from the thing that was first inserted into the popup stack, to the thing that was most recently inserted into the popup stack.

Author: Addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

@mfreed7, I thought I recalled that we were intending to work from most recently added popup in the stack (shown on top-most layer) to the least-recently added, which is a bit different. Is that the case with your prototype?

> 1. If _candidate subject_ is null, skip to step 6.

Reviewer 4: Like in coding, avoid "goto" in specs. Instead use "If/else". E.g. "If candidate subject is not null: ... nest 5 steps."

It's also not clear what skipping to a step which starts with "Otherwise" means.

Author: Addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

2. Otherwise, starting with the _candidate subject_, walk the
   [flat tree](https://drafts.csswg.org/css-scoping/#flattening) to determine whether the _subject_
   is an ancestor to the _candidate subject_. If this condition is met, then return.
3. Otherwise, if the _invoker_ is not null, starting with the _invoker_, walk the
   [flat tree](https://drafts.csswg.org/css-scoping/#flattening) to determine whether the _subject_
   is an ancestor to the _invoker_. If this condition is met, then return.
4. Otherwise, if the _anchor element_ is not null, starting with the _anchor element_, walk the
   [flat tree](https://drafts.csswg.org/css-scoping/#flattening) to determine whether the _subject_
   is an ancestor to the _anchor element_. If this condition is met, then return.
5. If _start node_ is not null, starting with the _start node_, walk the
   [flat tree](https://drafts.csswg.org/css-scoping/#flattening) to determine whether the _subject_

>    is the _start node_ or an ancestor to the _start node_. If either condition is met, then return.

Reviewer 4: Probably you don't need to do a full tree walk to find out if _subject_ is _start node_; just an equality check (in a separate step) would suffice.

Author: Addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

6. Otherwise, for this _subject_, run the

>    [hiding a `popup` element steps](#hiding-popup-steps), with _subject_ as _candidate subject_.

Reviewer 4: Do you want to pass along _invoker_?

Author: Actually yeah, even if it's just to set that to null. Addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

<h3 id="hiding-popup-steps">
  Hiding a <code>popup</code> element steps
</h3>

The **hiding a `popup` element steps**, given a `popup` element _candidate subject_ and an optional
element _invoker_, are as follows:

1. Remove the _candidate subject_ from the
   [node document](https://dom.spec.whatwg.org/#concept-node-document)'s [popup stack](#popup-stack).

> 2. Set the `open` IDL attribute on the _candidate subject_ to `false`.

Reviewer 4: As before, setting IDL attributes directly isn't great.

Author: Addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

3. Set the _candidate subject_'s _open state_ to `false`.

> 4. If applicable, empty the _candidate subject_'s _invoker_.

Reviewer 4: I think this intends something like "set _candidate subject_'s invoker to null"? (In which case no vague "If applicable" is needed.) Or is it intending to remove all DOM children of the invoker?

Author: I think I'd copied this style from elsewhere in the spec, but the null intent is correct. Addressed in 5fc5041409d28c395e81cf30f039a7c0e269b7fe

   on the [user interaction task source](https://html.spec.whatwg.org/multipage/webappapis.html#user-interaction-task-source)
   given the _candidate subject_ element to
   [fire an event](https://dom.spec.whatwg.org/#concept-event-fire)
   named `hide` at _candidate subject_.

---

## HTML Incubation Text: User Interaction > Focus > Processing Model

<p class="note">
  The following text to be inserted into the
  <a href="https://html.spec.whatwg.org/multipage/interaction.html#focus-processing-model)">
    get the focusable area steps
  </a>.
</p>

**If _focus target_ is a `popup` element**

1. If the `delegatesfocus` attribute is specified on _focus target_, return the first focusable
   descendent element of _focus target_, in
   [tree order](https://dom.spec.whatwg.org/#concept-tree-order), that is not inert.
2. If there is no such descendent, but `delegatesfocus` is specified, return the _focus target_.
3. Otherwise, if the `autofocus` attribute is specified on _focus target_, return the _focus target_.
4. Otherwise, if the `autofocus` attribute is specified on a descendent element of _focus target_,
   return the first such descendent element of _focus target_, in

>    [tree order](https://dom.spec.whatwg.org/#concept-tree-order), that is not inert.

Reviewer 4: Interesting. This allows autofocus to focus elements which are not normally focusable (as long as they are not inert). Is that intended behavior?

The normal autofocus behavior does not do this. It performs "getting the focusable area" on the element where autofocus is specified, which might return the element itself, or might return null. If it returns null, then the attribute does nothing.
